// fx-api/build/constants.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Stage = exports.Task = exports.VsCodeEnv = exports.Platform = exports.ProductName = exports.ConfigFolderName = void 0;
exports.ConfigFolderName = "fx";
exports.ProductName = "teamsfx";
var Platform;
(function (Platform) {
    Platform["VSCode"] = "vsc";
    Platform["VS"] = "vs";
    Platform["CLI"] = "cli";
})(Platform = exports.Platform || (exports.Platform = {}));
var VsCodeEnv;
(function (VsCodeEnv) {
    VsCodeEnv["local"] = "local";
    VsCodeEnv["codespaceBrowser"] = "codespaceBrowser";
    VsCodeEnv["codespaceVsCode"] = "codespaceVsCode";
    VsCodeEnv["remoteContainer"] = "remoteContainer";
})(VsCodeEnv = exports.VsCodeEnv || (exports.VsCodeEnv = {}));
var Task;
(function (Task) {
    Task["create"] = "create";
    Task["update"] = "update";
    Task["debug"] = "debug";
    Task["provision"] = "provision";
    Task["deploy"] = "deploy";
    Task["publish"] = "publish";
    Task["userTask"] = "userTask";
})(Task = exports.Task || (exports.Task = {}));
var Stage;
(function (Stage) {
    Stage["create"] = "create";
    Stage["update"] = "update";
    Stage["debug"] = "debug";
    Stage["provision"] = "provision";
    Stage["deploy"] = "deploy";
    Stage["publish"] = "publish";
    Stage["userTask"] = "userTask";
})(Stage = exports.Stage || (exports.Stage = {}));
