// fx-api/build/ci/aadManager.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AadManager = void 0;
const tslib_1 = require("tslib");
const axios_1 = tslib_1.__importDefault(require("axios"));
const mockGraphTokenProvider_1 = tslib_1.__importDefault(require("./mockGraphTokenProvider"));
function delay(ms) {
    // tslint:disable-next-line no-string-based-set-timeout
    return new Promise((resolve) => setTimeout(resolve, ms));
}
class AadManager {
    constructor() {
        AadManager.axios = undefined;
        AadManager.provider = undefined;
    }
    static async init(provider) {
        if (!AadManager.instance) {
            AadManager.instance = new AadManager();
        }
        if (AadManager.provider !== (provider || mockGraphTokenProvider_1.default)) {
            AadManager.provider = provider || mockGraphTokenProvider_1.default;
            const token = await AadManager.provider.getAccessToken();
            AadManager.axios = axios_1.default.create({
                baseURL: "https://graph.microsoft.com/v1.0/",
                headers: {
                    authorization: `Bearer ${token}`,
                    ConsistencyLevel: "eventual",
                    "content-type": "application/json",
                },
            });
        }
        return Promise.resolve(AadManager.instance);
    }
    async searchAliveAadApps(prefix, offsetHour = 0) {
        const result = await AadManager.axios.get(`applications?$filter=startswith(displayName, '${prefix}')&$count=true&$top=100&$orderby=displayName`);
        const apps = result.data.value;
        return Promise.resolve(apps.filter(app => Date.now() - new Date(app.createdDateTime).getTime() > offsetHour * 3600 * 1000));
    }
    async searchDeletedAadApps(prefix, offsetHour = 0) {
        const result = await AadManager.axios.get(`directory/deleteditems/microsoft.graph.application?$filter=startswith(displayName, '${prefix}')&$count=true&$top=100&$orderby=displayName`);
        const apps = result.data.value;
        return Promise.resolve(apps.filter(app => Date.now() - new Date(app.createdDateTime).getTime() > offsetHour * 3600 * 1000));
    }
    async searchAadApps(contain, offsetHour = 0) {
        return new Promise(async (resolve) => {
            const [aliveAadApps, deletedAadApps] = await Promise.all([
                this.searchAliveAadApps(contain, offsetHour),
                this.searchDeletedAadApps(contain, offsetHour)
            ]);
            return resolve(aliveAadApps.concat(deletedAadApps));
        });
    }
    async deleteAadAppById(id, retryTimes = 5) {
        return new Promise(async (resolve) => {
            try {
                await AadManager.axios.delete(`applications/${id}`);
            }
            finally {
                for (let i = 0; i < retryTimes; ++i) {
                    try {
                        await AadManager.axios.delete(`directory/deletedItems/${id}`);
                        return resolve(true);
                    }
                    catch (_a) {
                        await delay(2000);
                        if (i < retryTimes - 1) {
                            console.warn(`[Retry] clean up the Aad app failed with id: ${id}`);
                        }
                    }
                }
                return resolve(false);
            }
        });
    }
    async deleteAadApps(contain, offsetHour = 0, retryTimes = 5) {
        const aadApps = await this.searchAadApps(contain, offsetHour);
        console.log(`There are ${aadApps.length} Aad apps created ${offsetHour.toFixed(2)} hours ago. Deleting...`);
        const promises = aadApps.map(app => this.deleteAadAppById(app.id, retryTimes));
        const results = await Promise.all(promises);
        results.forEach((result, index) => {
            if (result) {
                console.log(`[Sucessfully] clean up the Aad app with id: ${aadApps[index].id}, appId: ${aadApps[index].appId}`);
            }
            else {
                console.log(`[Failed] no permission to clean up the Aad app with id: ${aadApps[index].id}, appId: ${aadApps[index].appId}`);
            }
        });
    }
}
exports.AadManager = AadManager;
