// fx-api/build/qm/question.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QTreeNode = exports.NodeType = void 0;
/**
 * reference:
 * https://www.w3schools.com/html/html_form_input_types.asp
 * https://www.w3schools.com/tags/att_option_value.asp
 */
var NodeType;
(function (NodeType) {
    NodeType["text"] = "text";
    NodeType["number"] = "number";
    NodeType["password"] = "password";
    NodeType["singleSelect"] = "singleSelect";
    NodeType["multiSelect"] = "multiSelect";
    NodeType["file"] = "file";
    NodeType["folder"] = "folder";
    NodeType["group"] = "group";
    NodeType["func"] = "func";
    NodeType["localFunc"] = "localFunc";
})(NodeType = exports.NodeType || (exports.NodeType = {}));
/**
 * QTreeNode is the tree node data structure, which have three main properties:
 * - data: data is either a group or question. Questions can be organized into a group, which has the same trigger condition.
 * - condition: trigger condition for this node to be activated;
 * - children: child questions that will be activated according their trigger condition.
 */
class QTreeNode {
    constructor(data) {
        this.data = data;
    }
    addChild(node) {
        if (!this.children) {
            this.children = [];
        }
        this.children.push(node);
        return this;
    }
    validate() {
        //1. validate the cycle depedency
        //2. validate the name uniqueness
        //3. validate the params of RPC
        // if (this.data.type === NodeType.group && (!this.children || this.children.length === 0)) return false;
        return true;
    }
    /**
     * trim the tree
     */
    trim() {
        if (this.children) {
            const newChildren = [];
            for (const node of this.children) {
                const trimed = node.trim();
                if (trimed)
                    newChildren.push(trimed);
            }
            this.children = newChildren;
        }
        if (this.data.type === NodeType.group) {
            if (!this.children || this.children.length === 0)
                return undefined;
            if (this.children.length === 1) {
                this.children[0].condition = this.condition;
                return this.children[0];
            }
        }
        return this;
    }
}
exports.QTreeNode = QTreeNode;
